crates/bevy_render/src/mesh/primitives/dim3/sphere.rs

```diff
@@ -0,0 +1,268 @@
+        /// The number of subdivisions used. 79 is the largest allowed value for a mesh to be generated.
+        subdivisions: usize,
+        /// The number of vertices generated. 65535 is the largest allowed value for a mesh to be generated.
+        number_of_resulting_points: usize,
+    },
+}
+
+/// A type of sphere mesh.
+#[derive(Clone, Copy, Debug)]
+pub enum SphereKind {
+    /// An icosphere, a spherical mesh that consists of equally sized triangles.
+    Ico {
+        /// The number of subdivisions applied.
+        /// The number of faces quadruples with each subdivision.
+        subdivisions: usize,
+    },
+    /// A UV sphere, a spherical mesh that consists of quadrilaterals
+    /// apart from triangles at the top and bottom.
+    Uv {
+        /// The number of longitudinal sectors, aka the horizontal resolution.
+        #[doc(alias = "horizontal_resolution")]
+        sectors: usize,
+        /// The number of latitudinal stacks, aka the vertical resolution.
+        #[doc(alias = "vertical_resolution")]
+        stacks: usize,
+    },
+}
+
+impl Default for SphereKind {
+    fn default() -> Self {
+        Self::Ico { subdivisions: 5 }
+    }
+}
+
+/// A builder used for creating a [`Mesh`] with an [`Sphere`] shape.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct SphereMeshBuilder {
+    /// The [`Sphere`] shape.
+    pub sphere: Sphere,
+    /// The type of sphere mesh that will be built.
+    pub kind: SphereKind,
+}
+
+impl SphereMeshBuilder {
+    /// Creates a new [`SphereMeshBuilder`] from a radius and [`SphereKind`].
+    #[inline]
+    pub const fn new(radius: f32, kind: SphereKind) -> Self {
+        Self {
+            sphere: Sphere { radius },
+            kind,
+        }
+    }
+
+    /// Sets the [`SphereKind`] that will be used for building the mesh.
+    #[inline]
+    pub const fn kind(mut self, kind: SphereKind) -> Self {
+        self.kind = kind;
+        self
+    }
+
+    /// Builds a [`Mesh`] according to the configuration in `self`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the sphere is a [`SphereKind::Ico`] with a subdivision count
+    /// that is greater than or equal to `80` because there will be too many vertices.
+    pub fn build(&self) -> Mesh {
+        match self.kind {
+            SphereKind::Ico { subdivisions } => self.ico(subdivisions).unwrap(),
+            SphereKind::Uv { sectors, stacks } => self.uv(sectors, stacks),
+        }
+    }
+
+    /// Creates an icosphere mesh with the given number of subdivisions.
+    ///
+    /// The number of faces quadruples with each subdivision.
+    /// If there are `80` or more subdivisions, the vertex count will be too large,
+    /// and an [`IcosphereError`] is returned.
+    ///
+    /// A good default is `5` subdivisions.
+    pub fn ico(&self, subdivisions: usize) -> Result<Mesh, IcosphereError> {
+        if subdivisions >= 80 {
+            /*
+            Number of triangles:
+            N = 20
```

Comment:
I imagine this and the following parameters N, E, V are not actually fixed but depend on the number of subdivisions?

Reply:
No; I believe these are how many triangles, edges, and vertices an icosphere with 1 subdivision would have, and then the mathematical functions below use these constants to determine how many vertices the subdivided version would have.

I didn't write this though, it's also copied from the "old" icosphere.